rockcraft.yaml

```diff
@@ -0,0 +1,203 @@
+    plugin: dump
+    source: https://github.com/discourse/discourse-solved.git
+    source-depth: 1
+    source-commit: d6c8089ca38611b09a8edb29d64f359bcef11f11
+    after:
+      - get-discourse
+    organize:
+      "*": srv/discourse/app/plugins/discourse-solved/
+  get-markdown-note-plugin-source:
+    plugin: dump
+    source: https://github.com/canonical-web-and-design/discourse-markdown-note.git
+    source-depth: 1
+    after:
+      - get-discourse
+    organize:
+      "*": srv/discourse/app/plugins/discourse-markdown-note/
+  get-mermaid-plugin-source:
+    plugin: dump
+    source: https://github.com/unfoldingWord-dev/discourse-mermaid.git
+    source-depth: 1
+    after:
+      - get-discourse
+    organize:
+      "*": srv/discourse/app/plugins/discourse-mermaid/
+  # get-saml-plugin-source:
+  #   plugin: dump
+  #   source: https://github.com/discourse/discourse-saml.git
+  #   source-depth: 1
+  #   after:
+  #     - get-discourse
+  #   organize:
+  #     "*": srv/discourse/app/plugins/discourse-saml/
+  #   build-packages:
+  #     - ruby2.7
+  #     - ruby2.7-dev
+  #   override-stage: |
+  #     craftctl default
+
+  #     set -x
+  #     mkdir -p "${CRAFT_STAGE}/srv/discourse/app/.gem"
+
+  #     mkdir -p bin
+  #     gem install -n "bin" bundler
+  #     gem install bundler
+  #     bin/bundle config set --local path 'bundle'
+  #     grep ^gem $CRAFT_STAGE/srv/discourse/app/plugins/discourse-saml/plugin.rb >> ${CRAFT_STAGE}/srv/discourse/app/plugins/discourse-saml/Gemfile
+  #     bin/bundle install --gemfile="${CRAFT_STAGE}/srv/discourse/app/plugins/discourse-saml/Gemfile"
+  #     # echo "finished installing bundles"
+  #     # gem install macaddr -v 1.0.0 -i /srv/discourse/app/plugins/discourse-saml/gems/2.7.0 --no-document --ignore-dependencies --no-user-install
+
+  #     # cat plugin.rb
+  #     # grep ^gem plugin.rb >> ${CRAFT_STAGE}/srv/discourse/app/discourse-saml/Gemfile
+  #     # ${CRAFT_STAGE}/srv/discourse/app/bin/bundle install --gemfile=${CRAFT_STAGE}/srv/discourse/app/discourse-saml/Gemfile --path=${CRAFT_STAGE}/srv/discourse/app/discourse-saml/gems
+  get-patches:
+    plugin: dump
+    source: image/patches
+    organize:
+      "*": srv/discourse/app/patches/
+  apply-patches:
+    plugin: nil
+    build-packages:
+      - git
+    after:
+      - get-discourse
+      - get-patches
+    override-stage: |
+      git -C srv/discourse/app apply patches/lp1903695.patch
+      git -C srv/discourse/app apply patches/anonymize_user.patch
+  build-discourse-assets:
+    plugin: nil
+    after:
+      - get-discourse
+    override-build: |
+      curl -fsSL https://deb.nodesource.com/setup_18.x | bash -
+      apt-get install -y nodejs
+      node --version
+      npm --version
+
+      npm install -g terser uglify-js yarn
+      yarn --cwd ${CRAFT_STAGE}/srv/discourse/app install --production --frozen-lockfile
+      yarn --cwd ${CRAFT_STAGE}/srv/discourse/app cache clean
+  get-gems:
```

Comment:
Is the `get-*` recommended naming? I think these are parts, i.e., components so perhaps just `gems` or `discourse` etc is enough?

Reply:
I like the idea that the name of the parts (with the `after` property) could give the whole story of the build.
For this rockfile it would be:
```
- install-node
- install-runtime-packages
- get-discourse (after install-node)
- get-solved-plugin-source (after get-discourse)
- get-markdown-note-plugin-source (after get-discourse)
- get-mermaid-plugin-source (after get-discourse)
- get-saml-plugin-source (after get-discourse)
- get-patches
- apply-patches (after get-patches and get-discourse)
- install-gems (after get-discourse, get-saml-plugin-source and apply-patches)
- get-scripts
```
I feel like it reads itself pretty well. Also parts are more "actions" than "things"

Reply:
The snapcraft documentation includes a definition of a part which states that it is a component: https://snapcraft.io/docs/glossary

Reply:
This is a good discussion on naming, I like it. The lines here do tell a story and I like it. Looking at the glossary I interpret it that a `part` is more of a component combined with instructions on how to fetch and build it and they are reusable, so I'm leaning more towards "actions" as well. If we treat them as components and name them as such, perhaps the parts above need to be combined to reflect the glossary and the idea that fetch and build is included in a component, so:

```text
- get-solved-plugin-source
- get-markdown-note-plugin-source
- get-mermaid-plugin-source
- ...
```
will result in something like
```text
- discourse-plugins
```
with the implication that they will be fetched and built in that part...
Just a thought on my end.

Reply:
any thoughts @gregory-schiano? Based on our interactions with the starcraft team and the docs it sounds like the intent for parts is not really actions (though they define actions to make a part available). So a part should be called `discourse` if it installs discourse and `dependencies` if it install dependencies etc